runtime/src/main/scala/akka/grpc/internal/Marshallers.scala

```diff
@@ -24,28 +25,22 @@ abstract class BaseMarshaller[T](val protobufSerializer: ProtobufSerializer[T])
         case _              => 32 * 1024
       })
 
-    // Blocking calls underneath...
-    // we can't avoid it for the moment because we are relying on the Netty's Channel API
-    val bytes = stream.read(buffer, 0, buffer.length) match {
-      case -1 =>
-        akka.util.ByteString.empty // EOF immediately
-      case n if n < buffer.length =>
-        // WARNING: buffer is retained in full below,
-        // which could be problematic if ProtobufSerializer.deserialize keeps a reference to the ByteString
-        akka.util.ByteString.fromArrayUnsafe(buffer, 0, n)
-      case full =>
-        val baos = new ByteArrayOutputStream(buffer.length * 2) // To avoid immediate resize
-        var bytesRead = full
-
-        do {
-          baos.write(buffer, 0, bytesRead)
-          bytesRead = stream.read(buffer)
-        } while (bytesRead >= 0)
-
-        akka.util.ByteString.fromArrayUnsafe(baos.toByteArray)
+    @tailrec def readBytes(buffer: Array[Byte], at: Int, result: akka.util.ByteString): akka.util.ByteString = {
+      val capacity = buffer.length - at
+      val bytesRead = stream.read(buffer, at, capacity)
+      if (bytesRead < 0) {
+          if (at > 0) result ++ akka.util.ByteString.fromArrayUnsafe(buffer, 0, at) // Potentially wasteful since at could be small
+          else result
+      } else {
+        // Reading 0 bytes from an EOF stream should still EOF (-1) but didn't so has more data
+        if (capacity == 0 && bytesRead == 0)
```

Comment:
so if `capacity == 0`, we call `read` just to find out if we're at EOF, but this way we don't need a single-byte `write` if there is more data after all.

Since we know by the contract of `read` that `capacity != 0 && bytesRead == 0` is invalid/impossible, I guess we don't strictly need the `capacity == 0` check here, so we could shave off something here - though that makes it even harder to follow for perhaps not too much gain.

This kinda starts to deserve a unittest, though ;)

Reply:
@raboof I put that check in there if any InputStream implementation doesn't exhibit this behavior, and as such avoiding to go into an endless loop or breaking out before reading everything.